src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -69,15 +75,29 @@ static void callback(ConstFSEventStreamRef streamRef,
     jobject watcherCallback = (jobject) clientCallBackInfo;
 
     for (int i = 0; i < numEvents; i++) {
-        // TODO[max] Lion has much more detailed flags we need accurately process. For now just reduce to SL events range.
-        FSEventStreamEventFlags flags = eventFlags[i] & 0xFF;
-        if ((flags & kFSEventStreamEventFlagMustScanSubDirs) != 0) {
-            reportEvent("RECDIRTY", paths[i], watcherCallback);
-        } else if (flags != kFSEventStreamEventFlagNone) {
-            reportEvent("RESET", NULL, watcherCallback);
+        FSEventStreamEventFlags flags = eventFlags[i];
+        printf("~~~~ Event flags: 0x%x for %s\n", flags, paths[i]);
+        const char *type;
+        if (IS_SET(flags, kFSEventStreamEventFlagMustScanSubDirs)) {
+            type = "INVALIDATE";
+        } else if (IS_SET(flags, kFSEventStreamEventFlagItemRenamed)) {
+            if (IS_SET(flags, kFSEventStreamEventFlagItemCreated)) {
+                type = "REMOVED";
+            } else {
+                type = "CREATED";
+            }
```

Comment:
This is really weird. For renames it seems that there are two events:

1) `kFSEventStreamEventFlagItemRenamed` + `kFSEventStreamEventFlagItemCreated` for the old path (i.e. the one that is not available anymore)
2) `kFSEventStreamEventFlagItemRenamed` only for the new path (i.e. that is now available).

For me the reverse would be more logical, and maybe using `kFSEventStreamEventFlagItemRemoved` for the old path would be even more logical. Mysterious!